integrations/cdi/oci/src/main/java/io/helidon/integrations/cdi/oci/AdpSelectionStrategy.java

```diff
@@ -0,0 +1,545 @@
+        Void produceBuilder(Selector selector, Config c, Annotation[] qualifiersArray) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override // AdpSelectionStrategy
+        ConfigFileAuthenticationDetailsProvider produce(Selector selector, Config c, Annotation[] qualifiersArray) {
+            return
+                this.produce(c.get("oci.config.path", String.class).orElse(null), // null on purpose; use OCI defaulting logic
+                             c.get("oci.config.profile", String.class)
+                             .orElse(c.get("oci.auth.profile", String.class)
+                                     .orElse("DEFAULT")));
+        }
+
+        private ConfigFileAuthenticationDetailsProvider produce(String ociConfigPath, String ociAuthProfile) {
+            try {
+                if (ociConfigPath == null) {
+                    // Don't get clever and try to use ~/.oci/config
+                    // as a default value; there is OCI-managed logic
+                    // to figure out the proper default; we want to
+                    // make sure it's used
+                    return new ConfigFileAuthenticationDetailsProvider(ociAuthProfile);
+                } else {
+                    return new ConfigFileAuthenticationDetailsProvider(ociConfigPath, ociAuthProfile);
+                }
+            } catch (FileNotFoundException fileNotFoundException) {
+                throw new UncheckedIOException(fileNotFoundException.getMessage(), fileNotFoundException);
+            } catch (NoSuchFileException noSuchFileException) {
+                throw new UncheckedIOException(noSuchFileException.getMessage(), noSuchFileException);
+            } catch (IOException ioException) {
+                // The underlying ConfigFileReader that does the real
+                // work does not throw a FileNotFoundException in this
+                // case (as it probably should).  We have no choice
+                // but to parse the error message.  See
+                // https://github.com/oracle/oci-java-sdk/blob/2.19.0/bmc-common/src/main/java/com/oracle/bmc/ConfigFileReader.java#L94-L98.
+                String message = ioException.getMessage();
+                if (message != null
+                    && message.startsWith("Can't load the default config from ")
+                    && message.endsWith(" because it does not exist or it is not a file.")) {
+                    throw new UncheckedIOException(message,
+                                                   (IOException) new FileNotFoundException(message).initCause(ioException));
+                }
+                throw new UncheckedIOException(message, ioException);
+            }
+        }
+    },
+
+    /**
+     * An {@link AdpSelectionStrategy} that works with {@link
+     * InstancePrincipalsAuthenticationDetailsProvider} instances.
+     */
+    INSTANCE_PRINCIPALS(InstancePrincipalsAuthenticationDetailsProvider.class,
```

Comment:
I actually spent a long time going through the various SDKs from other languages and the OCI documentation to see where the mistake lies: is it wrong to use "instance principal" or is it wrong to use "instance principals"?  By far, the plural form seems to be the most used in the documentation.  Having said that, maybe it's easiest to allow for both since clearly it is a point of confusion within the Oracle OCI SDK team and expanding outwards to the CLI team.

Reply:
See for example https://docs.oracle.com/en-us/iaas/Content/Identity/Tasks/callingservicesfrominstances.htm#ariaid-title3 but slang people seem to use the singular.  But if it's not officially documented that way….

Also there seems to have been a deliberate distinction that the SDK team was attempting to make between instance principals, plural, and resource principal, singular.  The plural-vs.-singular distinction shows up in, e.g., [`InstancePrincipalsAuthenticationDetailsProvider`](https://docs.oracle.com/en-us/iaas/tools/java/2.19.0/com/oracle/bmc/auth/InstancePrincipalsAuthenticationDetailsProvider.html) and [`ResourcePrincipalAuthenticationDetailsProvider`](https://docs.oracle.com/en-us/iaas/tools/java/2.19.0/com/oracle/bmc/auth/ResourcePrincipalAuthenticationDetailsProvider.html).

As another arbitrary example, the Python SDK has only plural forms:

![Screen Shot 2022-03-14 at 10 21 01 PM](https://user-images.githubusercontent.com/80764/158312150-1012058f-9a0b-40c2-9abb-27999cb6c380.png)

So I'm inclined to leave it the way it is, whether the `oci` command line uses the singular form or not.

Reply:
Ok I'm fine with this.